### PR TITLE
feat(autoware_adapi_v1_msgs): add msgs for MRM

### PR DESCRIPTION
**Pull request by mkuri**
## Description

- Add msg to express MRM status and behavior
- Add srv to activate each MRM

## Related links

https://github.com/autowarefoundation/autoware.universe/pull/2070
https://github.com/autowarefoundation/autoware.universe/issues/1856

## Tests performed

https://github.com/autowarefoundation/autoware.universe/pull/2070

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [x] The PR follows the [pull request guidelines].
- [x] The PR has been properly tested.
- [x] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Comment by isamu-takagi** [comment]
What should we do with acronyms? I personally prefer the names `MrmState` and `OperateMrm`.

File: autoware_adapi_v1_msgs/system/msg/MRMState.msg [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+uint16 UNKNOWN = 0
+
+uint16 NORMAL = 1
+uint16 MRM_OPERATING = 2
+uint16 MRM_SUCCEEDED = 3
+uint16 MRM_FAILED = 4
+
+uint16 NONE = 101
+uint16 EMERGENCY_STOP = 102
+uint16 COMFORTABLE_STOP = 103
```
Comment (kenji-miyake) [inline autoware_adapi_v1_msgs/system/msg/MRMState.msg:12]: I think it's better to add comments like Takagi-san did with other messages.
@isamu-takagi What do you think?

```suggestion
# For state
uint16 UNKNOWN = 0
uint16 NORMAL = 1
uint16 MRM_OPERATING = 2
uint16 MRM_SUCCEEDED = 3
uint16 MRM_FAILED = 4

# For behavior 
uint16 NONE = 101
uint16 EMERGENCY_STOP = 102
uint16 COMFORTABLE_STOP = 103
```

File: autoware_adapi_v1_msgs/system/msg/MRMState.msg [hunk 2]
```diff
@@ -0,0 +1,15 @@
+builtin_interfaces/Time stamp
+
+uint16 UNKNOWN = 0
+
+uint16 NORMAL = 1
+uint16 MRM_OPERATING = 2
+uint16 MRM_SUCCEEDED = 3
+uint16 MRM_FAILED = 4
+
+uint16 NONE = 101
```
Comment (kenji-miyake) [inline autoware_adapi_v1_msgs/system/msg/MRMState.msg:10]: What is the reason for using 3 digits for this?

File: autoware_adapi_v1_msgs/CMakeLists.txt [hunk 3]
```diff
@@ -22,6 +22,8 @@ rosidl_generate_interfaces(${PROJECT_NAME}
   planning/msg/SteeringFactorArray.msg
   planning/msg/VelocityFactor.msg
   planning/msg/VelocityFactorArray.msg
+  system/msg/MRMState.msg
+  system/srv/OperateMRM.srv
```
Comment (kenji-miyake) [inline autoware_adapi_v1_msgs/CMakeLists.txt:26]: https://github.com/orgs/autowarefoundation/discussions/2824#discussioncomment-3911938

```suggestion
  system/msg/MrmState.msg
  system/srv/OperateMrm.srv
```

File: autoware_adapi_v1_msgs/system/srv/OperateMRM.srv [hunk 4]
```diff
@@ -0,0 +1,4 @@
+builtin_interfaces/Time stamp
+bool operate
```
Comment (kenji-miyake) [inline autoware_adapi_v1_msgs/system/srv/OperateMRM.srv:2]: What does `operate` mean? Could you link the specification if it's written in the documentation?
cc @isamu-takagi

File: autoware_adapi_v1_msgs/system/msg/MRMState.msg (hunk 2 above)
Comment (mkuri) [reply autoware_adapi_v1_msgs/system/msg/MRMState.msg:11]: When you put a behavior variable in the state, it will work unintentionally. And vice versa.

Comment (kenji-miyake) [reply autoware_adapi_v1_msgs/system/msg/MRMState.msg:11]: Seeing the existing other messages, it seems they are not so.
https://github.com/autowarefoundation/autoware_adapi_msgs/blob/ac51d9cd88154dfc3d1de8ef8a0b3b4658bfee57/autoware_adapi_v1_msgs/planning/msg/SteeringFactor.msg

@isamu-takagi What do you think?

I think we should define the guidelines for this rule.

Comment (isamu-takagi) [reply autoware_adapi_v1_msgs/system/msg/MRMState.msg:11]: [My first idea](https://autowarefoundation.github.io/autoware-documentation/main/design/autoware-interfaces/#constants-and-enumeration) was as @mkuri -san said. However, as a result of implementation, I think the following is good.
- Group by definition
- Assign numbers to each group (allow duplication)
- Common group for shared constants (e.g. UNKNOWN, NONE)

File: autoware_adapi_v1_msgs/system/msg/MrmState.msg [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+uint16 UNKNOWN = 0
+uint16 NORMAL = 1
+uint16 MRM_OPERATING = 2
+uint16 MRM_SUCCEEDED = 3
+uint16 MRM_FAILED = 4
+
+# For behavior
+uint16 NONE = 101
+uint16 EMERGENCY_STOP = 102
+uint16 COMFORTABLE_STOP = 103
```
Comment (isamu-takagi) [inline autoware_adapi_v1_msgs/system/msg/MrmState.msg:13]:
```suggestion
# For common use
uint16 UNKNOWN = 0

# For state
uint16 NORMAL = 1
uint16 MRM_OPERATING = 2
uint16 MRM_SUCCEEDED = 3
uint16 MRM_FAILED = 4

# For behavior
uint16 NONE = 1
uint16 EMERGENCY_STOP = 2
uint16 COMFORTABLE_STOP = 3
```

File: autoware_adapi_v1_msgs/system/msg/MRMState.msg (hunk 2 above)
Comment (isamu-takagi) [reply autoware_adapi_v1_msgs/system/msg/MRMState.msg]: @mkuri @kenji-miyake 
Applying the above policy will look like this. https://github.com/autowarefoundation/autoware_adapi_msgs/pull/16#discussion_r1017582282

File: autoware_adapi_v1_msgs/system/srv/OperateMRM.srv (hunk 4 above)
Comment (isamu-takagi) [reply autoware_adapi_v1_msgs/system/srv/OperateMRM.srv]: It was planned to provide only the state. Would you like to add this API? @mkuri 
https://autowarefoundation.github.io/autoware-documentation/main/design/autoware-interfaces/ad-api/list/api/fail_safe/

And is this service for forcing to trigger MRM by applications? If so, what behavior of MRM is selected?

File: autoware_adapi_v1_msgs/system/msg/MRMState.msg (hunk 2 above)
Comment (mkuri) [reply autoware_adapi_v1_msgs/system/msg/MRMState.msg]: Fix constants with the following commit.
https://github.com/autowarefoundation/autoware_adapi_msgs/pull/16/commits/9f86c2814267f8c07a09597c7558e18615b4ef7e

**Review by isamu-takagi — APPROVED** [review]
LGTM

**Review by kenji-miyake — APPROVED** [review]

**Comment by mkuri** [comment]
Since the DCO could no longer be resolved due to the rebase, I would like to re-create a new PR.

**Comment by kenji-miyake** [comment]
@mkuri I've manually set the DCO check to pass. You don't need to create a new PR.

**Pull request merged**